File: http-servers/typescript/deno-oak/src/config/env.ts
import process from "node:process";
import { z } from "zod";

const zEnv = z.object({
  ENV: z.enum(["dev", "prod"]).default("dev"),
  HOST: z
    .union([z.ipv4().trim(), z.literal("localhost")])
    .transform((val) => (val === "localhost" ? "0.0.0.0" : val))
    .default("0.0.0.0"),
  PORT: z.coerce.number().int().min(1).max(65535).default(3004),
  POSTGRES_URL: z.string().trim().default(
    "postgres://postgres:postgres@localhost:5432/benchmarks"
  ),
  MONGODB_URL: z.string().trim().default("mongodb://localhost:27017"),
  MONGODB_DB: z.string().trim().default("benchmarks"),
  REDIS_URL: z.string().trim().default("redis://localhost:6379"),
  CASSANDRA_CONTACT_POINTS: z
    .string()
    .trim()
    .default("localhost")
    .transform((value) =>
      value
        .split(",")
        .map((item) => item.trim())
        .filter(Boolean)
    ),
  CASSANDRA_KEYSPACE: z.string().trim().default("benchmarks"),
  CASSANDRA_LOCAL_DATACENTER: z.string().trim().default("datacenter1")
});

export const env = zEnv.parse(process.env);

File: shared/kotlin/src/main/kotlin/com/bench/shared/Env.kt
package com.bench.shared

/**
 * Environment contract shared by every Kotlin server (mirrors `shared/rust/env`
 * and `shared/go/config`). Host/port are NOT here — Ktor's `EngineMain` reads
 * those from `application.yaml` (`$PORT`/`$HOST` overrides); this holds only the
 * infrastructure config the app reads directly.
 */
data class Env(
    val env: String,
    val postgresUrl: String,
    val mongodbUrl: String,
    val mongodbDb: String,
    val redisUrl: String,
    val cassandraContactPoints: List<String>,
    val cassandraLocalDatacenter: String,
    val cassandraKeyspace: String,
    val jwtSecret: String,
) {
    /** True in production mode — the caller disables request logging here. */
    val isProd: Boolean get() = env == "prod"

    companion object {
        fun load(): Env =
            Env(
                env = envOr("ENV", "dev"),
                postgresUrl = envOr("POSTGRES_URL", "postgres://postgres:postgres@localhost:5432/benchmarks"),
                mongodbUrl = envOr("MONGODB_URL", "mongodb://localhost:27017"),
                mongodbDb = envOr("MONGODB_DB", "benchmarks"),
                redisUrl = envOr("REDIS_URL", "redis://localhost:6379"),
                cassandraContactPoints =
                    envOr("CASSANDRA_CONTACT_POINTS", "localhost")
                        .split(",")
                        .map { it.trim() }
                        .filter { it.isNotEmpty() },
                cassandraLocalDatacenter = envOr("CASSANDRA_LOCAL_DATACENTER", "datacenter1"),
                cassandraKeyspace = envOr("CASSANDRA_KEYSPACE", "benchmarks"),
                // Shared HS256 secret for the web suite; dev default must match the
                // other languages' shared env modules and the contract harness.
                jwtSecret = envOr("JWT_SECRET", "benchmarks-shared-jwt-secret-dev-default"),
            )

        private fun envOr(
            key: String,
            default: String,
        ): String = System.getenv(key)?.takeIf { it.isNotEmpty() } ?: default
    }
}
